ocaml/parsing/parser.mly

```diff
@@ -858,7 +858,7 @@ let mkghost_newtype_function_body newtypes body_constraint body ~loc =
         let body = Exp.mk (mkexp_desc_constraint body type_constraint) ~loc in
         exp_with_modes loc mode_annotations body
   in
-  mk_newtypes ~loc newtypes wrapped_body
+  mk_newtypes ~loc:(ghost_loc loc) newtypes wrapped_body
```

Comment:
As discussed, I don't think the outermost node should be ghost if the user used the `fun` syntax.

Reply:
Fixed this by changing the `mk_newtypes` function.

It also affects the parsing of `let a : type a b c. unit = ()` which we do differently from upstream to support layout annotation. Now the outermost newtype expression is not marked as ghost compared to before.

We can't return just the `pexp_desc` like the upstream implementation due to layout annotation being encoded as attributes.

Reply:
To make sure I'm interpreting you right: the fact you mention (that the newtype node in `let a : type a b c. unit = ()` is not ghost) is not something introduced by this PR, right?

Reply:
My bad that part is not correct. To clearify, the parsing of `let a : type a b c. unit = ()` is already different from upstream before this PR. In our original implementation before this PR, It's parsed as:

```
Ptop_def
  [
    structure_item (//toplevel//[1,0+0]..[1,0+29])
      Pstr_value Nonrec
      [
        <def>
          pattern (//toplevel//[1,0+4]..[1,0+24]) ghost
            Ppat_constraint
            pattern (//toplevel//[1,0+4]..[1,0+5])
              Ppat_var "a" (//toplevel//[1,0+4]..[1,0+5])
            core_type (//toplevel//[1,0+20]..[1,0+24]) ghost
              Ptyp_poly 'a 'b 'c
              core_type (//toplevel//[1,0+20]..[1,0+24])
                Ptyp_constr "unit" (//toplevel//[1,0+20]..[1,0+24])
                []
          expression (//toplevel//[1,0+4]..[1,0+29])
            Pexp_newtype "a"
            expression (//toplevel//[1,0+4]..[1,0+29])
              Pexp_newtype "b"
              expression (//toplevel//[1,0+4]..[1,0+29])
                Pexp_newtype "c"
                expression (//toplevel//[1,0+4]..[1,0+29])
                  Pexp_constraint
                  expression (//toplevel//[1,0+27]..[1,0+29])
                    Pexp_construct "()" (//toplevel//[1,0+27]..[1,0+29])
                    None
                  core_type (//toplevel//[1,0+20]..[1,0+24])
                    Ptyp_constr "unit" (//toplevel//[1,0+20]..[1,0+24])
                    []
      ]
  ]
```

After this PR, it will be parsed as:

```
Ptop_def
  [
    structure_item (//toplevel//[1,0+0]..[1,0+29])
      Pstr_value Nonrec
      [
        <def>
          pattern (//toplevel//[1,0+4]..[1,0+24]) ghost
            Ppat_constraint
            pattern (//toplevel//[1,0+4]..[1,0+5])
              Ppat_var "a" (//toplevel//[1,0+4]..[1,0+5])
            core_type (//toplevel//[1,0+20]..[1,0+24]) ghost
              Ptyp_poly 'a 'b 'c
              core_type (//toplevel//[1,0+20]..[1,0+24])
                Ptyp_constr "unit" (//toplevel//[1,0+20]..[1,0+24])
                []
          expression (//toplevel//[1,0+4]..[1,0+29])
            Pexp_newtype "a"
            expression (//toplevel//[1,0+4]..[1,0+29]) ghost
              Pexp_newtype "b"
              expression (//toplevel//[1,0+4]..[1,0+29]) ghost
                Pexp_newtype "c"
                expression (//toplevel//[1,0+4]..[1,0+29])
                  Pexp_constraint
                  expression (//toplevel//[1,0+27]..[1,0+29])
                    Pexp_construct "()" (//toplevel//[1,0+27]..[1,0+29])
                    None
                  core_type (//toplevel//[1,0+20]..[1,0+24])
                    Ptyp_constr "unit" (//toplevel//[1,0+20]..[1,0+24])
                    []
      ]
  ]
```

Difference is that the original version doesn't have ghost on any of the `Pexp_newtype` expressions and now we put  ghost on all of them except the outermost one. This feels like an improvement to me. Let me know if you disagree.

Reply:
That sounds ok to me.